### ISSUE TITLE
Binding of code using `pybind11` objects.

### ISSUE THREAD

**Issue by zwimer** [RosettaCommons/binder#221]
TLDR: Allow functions using `pybind11::dict` without generating bindings for `pybind11::dict` since pybind11 does not need custom bindings for its own objects.

---

Consider:
```C++
namespace Test { inline std::string foo(pybind11::dict& d) { return pybind11::repr(d); } }
```

If I want to generate bindings for `+namespace Test` binder generates many files worth of code because it generates bindings for `pybind11::dict` and dependencies (the log is below). But this shouldn't need to be done since when compiling with the pybind11 headers pybind11 will know exactly what `pybind11::dict` is.

If I try to do `-class pybind11::dict` or something like that, the code for the `foo` function simply will not generate. Also, a lot of the code is put into `unknown/` rather than the `pybind11/` output directories, so skipping that directory is not sufficient for removing the unwanted code.

Is there a good way to either skip generating code for `pybind11::dict` or to simply remove it from the output directory after the fact?

---

```
Process input file /in/include.hpp
Read config from file /in/a.conf
Generate bindings, pass 1...
Binding: class std::basic_string<char> Test::foo(class pybind11::dict &)
Generate bindings, pass 2...
Binding: pybind11::dict
Generate bindings, pass 3...
Binding: pybind11::handle
Binding: pybind11::object
Binding: pybind11::detail::generic_iterator<pybind11::detail::iterator_policies::dict_readonly>
Binding: pybind11::detail::iterator_policies::dict_readonly
Binding: pybind11::str
Binding: pybind11::bytes
Generate bindings, pass 4...
Binding: _object
Binding: pybind11::detail::object_api<pybind11::handle>
Binding: pybind11::detail::args_proxy
Binding: pybind11::iterator
Generate bindings, pass 5...
Binding: pybind11::detail::accessor<pybind11::detail::accessor_policies::obj_attr>
Binding: pybind11::detail::accessor<pybind11::detail::accessor_policies::str_attr>
Binding: pybind11::detail::accessor<pybind11::detail::accessor_policies::generic_item>
Binding: pybind11::detail::pyobject_tag
Binding: pybind11::detail::object_api<pybind11::detail::accessor<pybind11::detail::accessor_policies::str_attr>>
Binding: pybind11::detail::object_api<pybind11::detail::accessor<pybind11::detail::accessor_policies::generic_item>>
Binding: pybind11::detail::object_api<pybind11::detail::accessor<pybind11::detail::accessor_policies::obj_attr>>
Binding: pybind11::detail::kwargs_proxy
Binding: pybind11::tuple
Binding: pybind11::cpp_function
Binding: pybind11::module_
Generate bindings, pass 6...
Binding: PyModuleDef
Binding: pybind11::detail::accessor<pybind11::detail::accessor_policies::tuple_item>
Binding: pybind11::detail::object_api<pybind11::detail::accessor<pybind11::detail::accessor_policies::tuple_item>>
Binding: pybind11::detail::generic_iterator<pybind11::detail::iterator_policies::sequence_fast_readonly>
Binding: pybind11::detail::iterator_policies::sequence_fast_readonly
Binding: pybind11::function
Generate bindings, pass 7...
Sorting Binders...
Sorting Binders... Done.
Writing code...
Writing /out/unknown/unknown.cpp
Writing /out/unknown/unknown_1.cpp
Writing /out/unknown/unknown_2.cpp
Writing /out/unknown/unknown_3.cpp
Writing /out/unknown/unknown_4.cpp
Writing /out/unknown/unknown_5.cpp
Writing /out/unknown/unknown_6.cpp
Writing /out/unknown/unknown_7.cpp
Writing /out/pybind11/pybind11.cpp
Writing /out/pybind11/pybind11_1.cpp
Writing /out/unknown/unknown_8.cpp
Writing /out/unknown/unknown_9.cpp
Writing /out/unknown/unknown_10.cpp
Writing /out/unknown/unknown_11.cpp
Writing /out/unknown/unknown_12.cpp
Writing /out/unknown/unknown_13.cpp
Writing code... Done.
```

**Comment by zwimer** [comment]
The above generally seems true for certain STL classes too, like `std::optional` which is handled by `#include <pybind11/stl.h>`.

**Comment by lyskov** [comment]
Thank you for letting me know @zwimer ! Yes, this is possible, - the best way to do it would be to add `pybind11::dict` as a python-builtin type. I just pushed the patch, please let me know if it works for you. Thanks,

**Comment by lyskov** [comment]
I also added `adding `std::optional` as python-builtin, - hope this helps!

**Comment by zwimer** [comment]
I'll probably make a PR adding a few other built-ins now that I know how. Thanks!

I'm linking this as a reference so that others who view this issue can see the `std::optional` commit.
https://github.com/RosettaCommons/binder/commit/0b3f581f8904d76495c8a755458eb6153fe0c09d

**Comment by zwimer** [comment]
@lyskov Why are are these lines commented out in the built-in list?
```
//"std::iterator", "std::reverse_iterator",
//"std::vector", "std::map", "std::list", "std::set",
```

When users `#include <pybind11/stl.h>` I believe these should be defined; that is where `std::optional` is defined too.

Is there a reason they have been commented out?

**Comment by lyskov** [comment]
Treating these types as Python-builtin might be or might be not a good approach depending on project needs. Since <pybind11/stl.h> approach is to generate new objects it might not be what someone want because of performance issues or if they need to be able to modify object-in-place. For example consider situation when project might have std::vector objects that is really large - converting them to Python lists would be a very expensive operations so alternative approach like providing bindings (instead of conversion) for such types would be needed. Please see https://pybind11.readthedocs.io/en/stable/advanced/cast/stl.html#binding-stl-containers for more details. Also if your project fall into later category you might find this useful: https://github.com/pybind/pybind11/blob/master/include/pybind11/stl_bind.h

**Comment by zwimer** [comment]
I made a PR to optionally enable / disable STL bindings, so users may use them if they please but do not need to. I also made a PR including pybind11 bindings

**Comment by zwimer** [comment]
Completed via https://github.com/RosettaCommons/binder/pull/224

Though https://github.com/RosettaCommons/binder/pull/230 would give more additional control.
